jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpTransportOverHTTP2.java

```diff
@@ -350,119 +351,363 @@ public void abort(Throwable failure)
-                commit = this.commit;
-                this.state = State.FAILED;
-                callback = this.callback;
-                this.callback = null;
-                this.failure = failure;
+                switch (_state)
+                {
+                    case SENDING:
+                    {
+                        _state = State.FAILING;
+                        _failure = failure;
+                        // Failing the callback will be done in postSend().
+                        return;
+                    }
+                    case IDLE:
+                    case PENDING:
+                    {
+                        _state = State.FAILED;
+                        _failure = failure;
+                        callback = _callback;
+                        commit = _commit;
+                        break;
+                    }
+                    default:
+                    {
+                        // This thread lost the race to fail the current send,
+                        // as other threads already succeeded or failed it.
+                        return;
+                    }
+                }
             }
-            if (LOG.isDebugEnabled())
-                LOG.debug(String.format("HTTP2 Response #%d/%h %s %s", stream.getId(), stream.getSession(),
-                    commit ? "commit" : "flush", callback == null ? "ignored" : "failed"), failure);
-            if (callback != null)
-                callback.failed(failure);
+            fail(callback, commit, failure);
         }
 
-        private boolean onIdleTimeout(Throwable failure)
+        private boolean idleTimeout(Throwable failure)
         {
-            boolean result;
-            Callback callback = null;
+            Callback callback;
+            boolean timeout;
             synchronized (this)
             {
-                // Ignore idle timeouts if not writing,
-                // as the application may be suspended.
-                result = state == State.WRITING;
-                if (result)
+                switch (_state)
                 {
-                    this.state = State.TIMEOUT;
-                    callback = this.callback;
-                    this.callback = null;
-                    this.failure = failure;
+                    case PENDING:
+                    {
+                        // The send was started but idle timed out, fail it.
+                        _state = State.FAILED;
+                        _failure = failure;
+                        callback = _callback;
+                        timeout = true;
+                        break;
+                    }
+                    case IDLE:
+                        // The application may be suspended, ignore the idle timeout.
+                    case SENDING:
+                        // A send has been started at the same time of an idle timeout;
+                        // Ignore the idle timeout and let the write continue normally.
```

Comment:
I kind of think the timeout should win this race.  Who knows what else has been closed already by the timeout, plus the timeout might be due to a really slow generation of the frame.

Reply:
Nope, we should support ignoring idle timeouts in case the application is idle as per Servlet specification, so if we were just unlucky here we should ignore the idle timeout.
I think we should enforce the idle timeout only when we are PENDING, i.e. really sure that we have initiated a write but it won't complete in time.